File: rust/python/mosqito_rs/roughness_dw.py
"""Daniel & Weber (1997) roughness.

Wraps the compiled bindings in :mod:`mosqito_rs._core` to match
``mosqito.sq_metrics.roughness.roughness_dw``'s exact public signatures.
"""

from __future__ import annotations

import numpy as np

from . import _core

__all__ = ["roughness_dw", "roughness_dw_freq"]


def roughness_dw(
    signal: np.ndarray, fs: float, overlap: float = 0.5
) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """Compute the roughness according to the Daniel and Weber method from a
    time signal.

    Matches :func:`mosqito.sq_metrics.roughness_dw`.

    Parameters
    ----------
    signal : numpy.ndarray
        Input time signal [Pa].
    fs : float
        Sampling frequency [Hz].
    overlap : float, default 0.5
        Overlapping coefficient for the 200 ms time windows.

    Returns
    -------
    R : numpy.ndarray
        Roughness value [asper], shape ``(Ntime,)``.
    R_spec : numpy.ndarray
        Specific roughness over the Bark axis, shape ``(47, Ntime)``.
    bark_axis : numpy.ndarray
        Frequency axis [Bark], shape ``(47,)``.
    time : numpy.ndarray
        Time axis [s], shape ``(Ntime,)``.
    """
    signal = np.ascontiguousarray(signal, dtype=np.float64)
    return _core.roughness_dw(signal, float(fs), float(overlap))


def roughness_dw_freq(
    spectrum: np.ndarray, freqs: np.ndarray
) -> tuple[float, np.ndarray, np.ndarray]:
    """Compute the roughness according to the Daniel and Weber method from a
    fine-band spectrum.

    Matches :func:`mosqito.sq_metrics.roughness_dw_freq` for a 1-D spectrum;
    the 2-D (multi-segment) case has not been ported.

    Parameters
    ----------
    spectrum : numpy.ndarray
        Input amplitude spectrum, shape ``(nperseg,)``.
    freqs : numpy.ndarray
        Input frequency axis [Hz], shape ``(nperseg,)``.

    Returns
    -------
    R : float
        Roughness value [asper].
    R_spec : numpy.ndarray
        Specific roughness over the Bark axis, shape ``(47,)``.
    bark_axis : numpy.ndarray
        Frequency axis [Bark], shape ``(47,)``.

    Warning
    -------
    The input spectrum must be an amplitude spectrum (use ``abs()`` on a
    complex spectrum).
    """
    spectrum = np.ascontiguousarray(spectrum, dtype=np.float64)
    freqs = np.ascontiguousarray(freqs, dtype=np.float64)
    return _core.roughness_dw_freq(spectrum, freqs)

File: rust/crates/mosqito-core/src/roughness/dw/roughness_dw.rs
//! Top-level Daniel & Weber roughness entry points, matching
//! `roughness_dw.py`/`roughness_dw_freq.py`.

use ndarray::Array2;
use num_complex::Complex64;
use rayon::prelude::*;

use super::gzi_weighting::gzi_weighting;
use super::h_weighting::h_weighting;
use super::main_calc::{
    channel_centres, roughness_dw_main_calc, roughness_dw_main_calc_with_setup, RoughnessDwSetup,
};
use crate::slm::{comp_spectrum_complex, SpectrumWindow};
use crate::utils::time_segmentation;

const N_CHANNEL: usize = 47;

/// `(R, R_spec, bark_axis, time_axis)` — [`roughness_dw`]'s result.
type RoughnessDwSegResult = (Vec<f64>, Array2<f64>, [f64; N_CHANNEL], Vec<f64>);

/// Computes Daniel & Weber roughness from a time signal, matching
/// `roughness_dw(signal, fs, overlap)`.
///
/// The signal is segmented into 200 ms blocks (matching MoSQITo's fixed
/// time resolution for this method), each block's roughness computed
/// independently and in parallel — mirroring `roughness_ecma`'s and
/// `loudness_zwst_perseg`'s use of rayon across segments.
///
/// # Panics
/// Panics if `signal` is shorter than one 200 ms block at `fs`.
pub fn roughness_dw(signal: &[f64], fs: f64, overlap: f64) -> RoughnessDwSegResult {
    let nperseg = (0.2 * fs) as usize;
    let noverlap = (overlap * nperseg as f64) as usize;
    let (blocks, time) = time_segmentation(signal, fs, nperseg, Some(noverlap));
    let nseg = blocks.ncols();

    let (spec, freq_axis) = comp_spectrum_complex(blocks.view(), fs, SpectrumWindow::Blackman);

    let h_weight = h_weighting(nperseg, fs);
    let zi = channel_centres();
    let gzi = gzi_weighting(&zi);
    // Every segment shares this signal's freq_axis/fs/block length, so the
    // ear-filter/threshold tables and FFT plans are identical across all
    // `nseg` calls below — built once here instead of once per segment.
    // See `RoughnessDwSetup`'s own docs.
    let setup = RoughnessDwSetup::new(&freq_axis, fs);

    let results: Vec<(f64, [f64; N_CHANNEL], [f64; N_CHANNEL])> = (0..nseg)
        .into_par_iter()
        .map(|col| {
            let spec_col: Vec<Complex64> = spec.column(col).to_vec();
            roughness_dw_main_calc_with_setup(&setup, &spec_col, &freq_axis, &gzi, &h_weight)
        })
        .collect();

    let r: Vec<f64> = results.iter().map(|&(r, _, _)| r).collect();
    let mut r_spec = Array2::<f64>::zeros((N_CHANNEL, nseg));
    for (col, &(_, spec_col, _)) in results.iter().enumerate() {
        for (z, &v) in spec_col.iter().enumerate() {
            r_spec[[z, col]] = v;
        }
    }
    let bark_axis = results
        .first()
        .map(|&(_, _, b)| b)
        .unwrap_or_else(channel_centres);

    (r, r_spec, bark_axis, time)
}

/// Computes Daniel & Weber roughness from a fine-band amplitude spectrum,
/// matching `roughness_dw_freq(spectrum, freqs)` for a 1-D spectrum (the
/// 2-D per-segment case is not ported — no caller in this crate needs it;
/// same scope-narrowing as `loudness_zwst_freq`/`noct_synthesis`).
///
/// `spectrum` must be a non-negative amplitude spectrum (not complex — use
/// `.norm()` first if starting from a complex FFT output).
///
/// # Panics
/// Panics if `spectrum` and `freqs` differ in length, or if `freqs` has
/// fewer than 2 points (needed to recover the implied sampling rate).
pub fn roughness_dw_freq(
    spectrum: &[f64],
    freqs: &[f64],
) -> (f64, [f64; N_CHANNEL], [f64; N_CHANNEL]) {
    assert_eq!(
        spectrum.len(),
        freqs.len(),
        "spectrum and freqs must have the same length"
    );
    assert!(freqs.len() >= 2, "freqs must have at least 2 points");

    let nperseg = spectrum.len();
    let df: f64 = freqs.windows(2).map(|w| w[1] - w[0]).sum::<f64>() / (freqs.len() - 1) as f64;
    let fs = (2.0 * nperseg as f64 * df).trunc();

    let h_weight = h_weighting(2 * nperseg, fs);
    let zi = channel_centres();
    let gzi = gzi_weighting(&zi);

    let spec_complex: Vec<Complex64> = spectrum.iter().map(|&s| Complex64::new(s, 0.0)).collect();
    roughness_dw_main_calc(&spec_complex, freqs, fs, &gzi, &h_weight)
}
